Plottables: add several new elliptical curve and area plot types

CoderPM2011 opened the pull request:
resolve #4721, #4653

- `Rotation` and change type to `Angle`
- Synchronized rotation with the angular direction of the Cartesian coordinate system  
  same as `PolarAxis` in #4321
- Drawing
  - Fixed `DrawArc()` distortion on ellipse
  - Added `DrawSector()`
  - Added `FillSector()`
- Enhanced Ellipse plot
  - Added `IsSector` property
  - Added `StartAngle` property
  - Added `SweepAngle` property
- Added new plot
  - `Arc`
  - `CircleSector`
  - `EllipticalArc`
  - `EllipticalSector`

## demo

### Arc
![image](https://github.com/user-attachments/assets/1afbf2ea-a308-4e51-8734-e3900543c72a)

```cs
formsPlot1.Plot.Axes.SquareUnits();
var sector = formsPlot1.Plot.Add.Arc(
    0, 0, 100,
    startAngle: Angle.FromDegrees(0),
    sweepAngle: Angle.FromDegrees(90));
sector.LineWidth = 3;
sector.LineColor = Colors.Black;
```

### CircleSector
![image](https://github.com/user-attachments/assets/85f77812-b26a-45cc-9994-c0dbcbfbd206)

```cs
formsPlot1.Plot.Axes.SquareUnits();
var sector = formsPlot1.Plot.Add.CircleSector(
    0, 0, 100,
    startAngle: Angle.FromDegrees(0),
    sweepAngle: Angle.FromDegrees(90));
sector.Rotation = Angle.FromDegrees(30);
sector.LineWidth = 3;
sector.LineColor = Colors.Black;
sector.FillColor = Colors.Blue.WithAlpha(0.5);
```
### EllipticalArc
![EllipticalArc](https://github.com/user-attachments/assets/6ed45a31-bbc1-478f-bc76-1725d961af37)

``` cs
private Angle sweep = Angle.FromDegrees(0);
private void timer1_Tick(object sender, EventArgs e)
{
    formsPlot1.Plot.Clear();
    var sector = formsPlot1.Plot.Add.EllipticalArc(
        0, 0, 100, 20,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: sweep);
    sector.Rotation = Angle.FromDegrees(0);
    sector.LineWidth = 3;
    sector.LineColor = Colors.Black;
    sector.FillColor = Colors.Blue.WithAlpha(0.5);

    formsPlot1.Plot.Add.Text($"{sweep}", 0, 0);
    formsPlot1.Refresh();

    sweep += Angle.FromDegrees(10);
    sweep = sweep.Normalized;
}
```
### EllipticalSector
![EllipticalSector](https://github.com/user-attachments/assets/79a594a2-1231-4309-8655-7f10ea245b30)

``` cs
private Angle sweep = Angle.FromDegrees(0);
private void timer1_Tick(object sender, EventArgs e)
{
    formsPlot1.Plot.Clear();
    var sector = formsPlot1.Plot.Add.EllipticalSector(
        0, 0, 100, 20,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: sweep);
    sector.Rotation = Angle.FromDegrees(30);
    sector.LineWidth = 3;
    sector.LineColor = Colors.Black;
    sector.FillColor = Colors.Blue.WithAlpha(0.5);

    formsPlot1.Plot.Add.Text($"{sweep}", 50, 0);
    formsPlot1.Refresh();

    sweep += Angle.FromDegrees(10);
    sweep = sweep.Normalized;
}
```

heartacker commented:
![Scribble_2025-01-26_06_35_51_595.jpg](https://github.com/user-attachments/assets/d7ebceae-8331-4827-aa4b-cabfb92e6a3f)

CoderPM2011 commented:
## New part
- Drawing
  - Added `DrawAnnularSector()` and `FillAnnularSector()`
  - Added `DrawEllipticalAnnulus()` and `FillEllipticalAnnulus()`
- Enhanced Ellipse plot
  - Added `InnerRadiusX` property
  - Added `InnerRadiusY` property
  - Added `IsAnnulus` property
- Added new plot
  - `AnnularEllipticalSector`
  - `AnnularSector`

##  Demo
![Demo](https://github.com/user-attachments/assets/de52b965-2362-48d9-85a3-ab03fec17bdb)

``` cs
private Angle rotation = Angle.FromDegrees(0);
private Angle sweep = Angle.FromDegrees(0);
private void timer1_Tick(object sender, EventArgs e)
{
    formsPlot1.Plot.Clear();
    var arc = formsPlot1.Plot.Add.Arc(
        new(-100, -100),
        90,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: sweep);
    arc.Rotation = rotation;
    arc.LineWidth = 3;
    arc.LineColor = Colors.Purple;

    var circleSector = formsPlot1.Plot.Add.CircleSector(
        new(-100, 100),
        90,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: -sweep);
    circleSector.Rotation = -rotation;
    circleSector.LineWidth = 3;
    circleSector.LineColor = Colors.Black;
    circleSector.FillColor = Colors.Green.WithAlpha(0.5);

    var annularSector = formsPlot1.Plot.Add.AnnularSector(
        new(100, 100),
        90, 20,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: -sweep);
    annularSector.Rotation = rotation;
    annularSector.LineWidth = 3;
    annularSector.LineColor = Colors.Black;
    annularSector.FillColor = Colors.Red.WithAlpha(0.5);

    var annularEllipticalSector = formsPlot1.Plot.Add.AnnularEllipticalSector(
        new(100, -100),
        90, 20, 50, 10,
        startAngle: Angle.FromDegrees(0),
        sweepAngle: sweep);
    annularEllipticalSector.Rotation = -rotation;
    annularEllipticalSector.LineWidth = 3;
    annularEllipticalSector.LineColor = Colors.Black;
    annularEllipticalSector.FillColor = Colors.Blue.WithAlpha(0.5);

    formsPlot1.Plot.Add.Text($"rotation[{rotation}], sweep[{sweep}]", -230, -200);
    formsPlot1.Refresh();

    rotation += Angle.FromDegrees(5);
    rotation = rotation.Normalized;
    sweep += Angle.FromDegrees(10);
    sweep = sweep.Normalized;
}
```

swharden commented:
Hi @CoderPM2011, wow a ton of work went into this! ✨ Thanks so much! I look forward to reviewing it now and merging it in so it can get into today's release 🚀 

Thanks @heartacker for the suggestions along the way 😄

swharden commented:
Hi @CoderPM2011, I'm going to go ahead and merge this in shortly, but is this the expected behavior? 🤔 

```cs
Coordinates center = new(0, 0);
double radius = 1.0;
Angle start = Angle.FromDegrees(-45);
Angle end = Angle.FromDegrees(90);
myPlot.Add.Arc(center, radius, start, end);
```

![image](https://github.com/user-attachments/assets/7d13c1f5-4dc7-47cd-86c8-b65ea6114e0c)

swharden commented:
Just an idea: arcs could have an option to define the start position (whether it's to the right or at the top) and direction (counter-clockwise or clockwise)

swharden commented:
Just an idea: arcs could have properties to return Coordinates of their start and end position. This would be useful for placing markers at the edges, or even arrowheads. Arrowheads would probably require knowing the rotation at the end of the line too 🤔

swharden commented:
I'm still going to merge so this can be released and fixed soon, but I think I found a bug

```cs
Coordinates center = new(0, 0);
double outerRadius = 2.0;
double innerRadius = 1.0;
Angle start = Angle.FromDegrees(45);
Angle sweep = Angle.FromDegrees(135);

var cs = myPlot.Add.AnnularSector(center, outerRadius, innerRadius, start, sweep);
cs.FillColor = Colors.Blue.WithAlpha(.2);
cs.LineColor = Colors.Black;
cs.LineWidth = 5;
```

![image](https://github.com/user-attachments/assets/692e39fe-c3a9-4eae-9ac6-ad20cc9b5e30)

swharden commented:
one more thing ... I didn't realize this before the merge, but I think part of this PR broke curves used by the Smith Chart because now it looks like this

![image](https://github.com/user-attachments/assets/47a14051-13fc-46e9-8ace-771b8e7e55a3)

swharden commented:
> Smith Chart

I found the break is due to the "correction" taking place inside `Drawing.DrawArc()`

swharden commented:
@CoderPM2011 I followed this PR up with #4742

On one hand it fixes the Smith Chart so I feel good about merging/publishing today, but on the other hand I suspect you put that features in there for a reason so I'd love your feedback about how you recommend we proceed. 

Thanks!
